controllers/jujucluster_controller.go

```diff
@@ -762,3 +839,173 @@ func createModel(ctx context.Context, jujuCluster *infrastructurev1beta1.JujuClu
+	if err != nil {
+		return nil, err
+	}
+
+	return address, nil
+}
+
+func createIntegrationsIfNeeded(ctx context.Context, jujuCluster *infrastructurev1beta1.JujuCluster, client *juju.Client, modelUUID string) (bool, error) {
+	log := log.FromContext(ctx)
+	created := false
+	integrationInputs := []juju.IntegrationInput{
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubeapi-load-balancer:certificates", "easyrsa:client"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-control-plane:certificates", "easyrsa:client"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-worker:certificates", "easyrsa:client"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"etcd:certificates", "easyrsa:client"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-control-plane:etcd", "etcd:db"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-control-plane:kube-control", "kubernetes-worker:kube-control"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-control-plane:loadbalancer-external", "kubeapi-load-balancer:lb-consumers"},
+		},
+		{
+			ModelUUID: modelUUID,
+			Endpoints: []string{"kubernetes-control-plane:loadbalancer-internal", "kubeapi-load-balancer:lb-consumers"},
+		},
+	}
+
+	for _, integrationInput := range integrationInputs {
+		integrationExists, err := client.Integrations.IntegrationExists(ctx, &integrationInput)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("Error querying existing integrations for integration: %+v", integrationInput))
+			return created, err
+		}
+		if !integrationExists {
+			log.Info("integration not found, creating it now", "integration", integrationInput)
+			response, err := client.Integrations.CreateIntegration(ctx, &integrationInput)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("error creating integration %+v", integrationInput))
+				return created, err
+			}
+			log.Info("created integration", "integration", integrationInput, "response", response)
+			created = true
+		}
+	}
+
+	return created, nil
+}
+
+func createApplicationsIfNeeded(ctx context.Context, jujuCluster *infrastructurev1beta1.JujuCluster, client *juju.Client, modelUUID string) (bool, error) {
+	log := log.FromContext(ctx)
+	created := false
+	easyRSACons, err := constraints.Parse("cores=1", "mem=4G", "root-disk=16G")
+	if err != nil {
+		log.Error(err, "error creating machine constraints")
+		return false, err
+	}
+
+	lbCons, err := constraints.Parse("cores=1", "mem=4G", "root-disk=16G")
+	if err != nil {
+		log.Error(err, "error creating machine constraints")
+		return created, err
+	}
+
+	createInputs := []juju.CreateApplicationInput{
+		{
+			ApplicationName: "easyrsa",
+			ModelUUID:       modelUUID,
+			CharmName:       "easyrsa",
+			CharmChannel:    "1.26/stable",
+			CharmBase:       "ubuntu@22.04",
+			CharmRevision:   32,
```

Comment:
It shouldnt be needed, but I just tried it and it looks like the charm lookup code does not seem to be working when revision is empty. Its probably something to do with the base. The terraform provider I based most of this on was written when series was still the way things were done (series has since been deprecated in favor of base). I will try and get some help from juju tomorrow and see if they can point me to a proper way of getting a revision when its not present.

Reply:
Alright the charm resolution does work as expected.  Previously the code was expecting a -1 to mean "unspecified" but I had mistakenly just deleted the revision completely from the struct initializer when I tested this. What happened then was that I was getting a 0 as a default value of the revision since the field is an int which lead to errors in resolving the charm using a revision 0. I have changed the int field to an int pointer so that the default value is nil, and the behavior is more in line with what would be expected, and its no longer necessary to specify -1, you can just leave the revision out completely.